Deployment of new version of application with OpenShift extension fails with `LabelSelectorRequirement(nil)}: field is immutable`

### Describe the bug

Hey @**Ioannis Canellos|226065** , I noticed something odd lately when deploying with the OpenShift extension... Opening as a bug since I'm not quite sure, but feel free to requalify.

It seems that the deployment fails because a selector includes the app version (the git SHA in this case), and selectors are immutable:

```
Error: [ERROR] 	[error]: Build step io.quarkus.kubernetes.deployment.KubernetesDeployer#deploy threw an exception: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: ***/apis/apps/v1/namespaces/prod-quarkus-github-lottery/deployments/quarkus-github-lottery. Message: Deployment.apps "quarkus-github-lottery" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.selector, message=Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable, reason=FieldValueInvalid, additionalProperties={})], group=apps, kind=Deployment, name=quarkus-github-lottery, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Deployment.apps "quarkus-github-lottery" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
```

I don't think I personally changed anything in this area in my app, so... could it be related to the switch from DeploymentConfig to Deployment?

More importantly, should this be addressed/documented in the extension, since it's apparently the default behavior?

See https://github.com/quarkusio/quarkus-github-lottery/actions/runs/8142539385/job/22252314385

### Expected behavior

Deployment should erase the previous version of the application -- at least it used to.

### Actual behavior

Deployment fails:

```
Error: [ERROR] 	[error]: Build step io.quarkus.kubernetes.deployment.KubernetesDeployer#deploy threw an exception: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PUT at: ***/apis/apps/v1/namespaces/prod-quarkus-github-lottery/deployments/quarkus-github-lottery. Message: Deployment.apps "quarkus-github-lottery" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=spec.selector, message=Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable, reason=FieldValueInvalid, additionalProperties={})], group=apps, kind=Deployment, name=quarkus-github-lottery, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Deployment.apps "quarkus-github-lottery" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/name":"quarkus-github-lottery", "app.kubernetes.io/version":"7df7e9a331fb67fdfe434ec56b49a3f52a944674"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
```

### How to Reproduce?

Deploy a new version of an application without removing the previous version one.

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.8.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/Constants.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/Constants.java
@@ -33,6 +33,8 @@ public final class Constants {
     public static final String ROUTE = "Route";
     public static final String ROUTE_API_GROUP = "route.openshift.io/v1";
 
+    static final String VERSION_LABEL = "app.kubernetes.io/version";
+
     static final String OPENSHIFT_APP_RUNTIME = "app.openshift.io/runtime";
     static final String S2I = "s2i";
     static final String DEFAULT_S2I_IMAGE_NAME = "s2i-java"; //refers to the Dekorate default image.

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
@@ -5,6 +5,7 @@ import static io.quarkus.kubernetes.deployment.Constants.KIND;
 import static io.quarkus.kubernetes.deployment.Constants.KNATIVE;
 import static io.quarkus.kubernetes.deployment.Constants.KUBERNETES;
 import static io.quarkus.kubernetes.deployment.Constants.MINIKUBE;
+import static io.quarkus.kubernetes.deployment.Constants.VERSION_LABEL;
 
 import java.io.File;
 import java.io.FileInputStream;
@@ -17,17 +18,22 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 import org.jboss.logging.Logger;
 
 import io.dekorate.utils.Serialization;
+import io.fabric8.kubernetes.api.builder.Visitor;
 import io.fabric8.kubernetes.api.model.APIResourceList;
 import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.KubernetesList;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.kubernetes.api.model.LabelSelectorFluent;
 import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
 import io.fabric8.kubernetes.api.model.batch.v1.Job;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClientException;
@@ -213,6 +219,11 @@ public class KubernetesDeployer {
             }
 
             list.getItems().stream().filter(distinctByResourceKey()).forEach(i -> {
+                Optional<HasMetadata> existing = Optional.ofNullable(client.resource(i).get());
+                checkLabelSelectorVersions(deploymentTarget, i, existing);
+            });
+
+            list.getItems().stream().filter(distinctByResourceKey()).forEach(i -> {
                 deployResource(deploymentTarget, client, i, optionalResourceDefinitions);
                 log.info("Applied: " + i.getKind() + " " + i.getMetadata().getName() + ".");
             });
@@ -239,6 +250,7 @@ public class KubernetesDeployer {
     private void deployResource(DeploymentTargetEntry deploymentTarget, KubernetesClient client, HasMetadata metadata,
             List<KubernetesOptionalResourceDefinitionBuildItem> optionalResourceDefinitions) {
         var r = findResource(client, metadata);
+        Optional<HasMetadata> existing = Optional.ofNullable(client.resource(metadata).get());
         if (shouldDeleteExisting(deploymentTarget, metadata)) {
             deleteResource(metadata, r);
         }
@@ -385,7 +397,6 @@ public class KubernetesDeployer {
         if (deploymentTarget.getDeployStrategy() != DeployStrategy.CreateOrUpdate) {
             return false;
         }
-
         return KNATIVE.equalsIgnoreCase(deploymentTarget.getName())
                 || resource instanceof Service
                 || (Objects.equals("v1", resource.getApiVersion()) && Objects.equals("Service", resource.getKind()))
@@ -397,5 +408,45 @@ public class KubernetesDeployer {
         Map<Object, Boolean> seen = new ConcurrentHashMap<>();
         return t -> seen.putIfAbsent(t.getApiVersion() + "/" + t.getKind() + ":" + t.getMetadata().getName(),
                 Boolean.TRUE) == null;
+    }
+
+    private static void checkLabelSelectorVersions(DeploymentTargetEntry deploymnetTarget, HasMetadata resource,
+            Optional<HasMetadata> existing) {
+        if (!existing.isPresent()) {
+            return;
+        }
+
+        if (resource instanceof Deployment) {
+            Optional<String> version = getLabelSelectorVersion(resource);
+            Optional<String> existingVersion = getLabelSelectorVersion(existing.get());
+            if (version.isPresent() && existingVersion.isPresent()) {
+                if (!version.get().equals(existingVersion.get())) {
+                    throw new IllegalStateException(String.format(
+                            "A previous Deployment with a conflicting label %s=%s was found in the label selector (current is %s=%s). As the label selector is immutable, you need to either align versions or manually delete previous deployment.",
+                            VERSION_LABEL, existingVersion.get(), VERSION_LABEL, version.get()));
+                }
+            } else if (version.isPresent()) {
+                throw new IllegalStateException(String.format(
+                        "A Deployment with a conflicting label %s=%s was in the label selector was requested (previous had no such label). As the label selector is immutable, you need to either manually delete previous deployment, or remove the label (consider using quarkus.%s.add-version-to-label-selectors=false).",
+                        VERSION_LABEL, version.get(), deploymnetTarget.getName().toLowerCase()));
+            } else if (existingVersion.isPresent()) {
+                throw new IllegalStateException(String.format(
+                        "A Deployment with no label in the label selector was requested (previous includes %s=%s). As the label selector is immutable, you need to either manually delete previous deployment, or ensure the %s label is present (consider using quarkus.%s.add-version-to-label-selectors=true).",
+                        VERSION_LABEL, existingVersion.get(), VERSION_LABEL, deploymnetTarget.getName().toLowerCase()));
+            }
+        }
+    }
+
+    private static Optional<String> getLabelSelectorVersion(HasMetadata resource) {
+        AtomicReference<String> version = new AtomicReference<>();
+        KubernetesList list = new KubernetesListBuilder().addToItems(resource).accept(new Visitor<LabelSelectorFluent<?>>() {
+            @Override
+            public void visit(LabelSelectorFluent<?> item) {
+                if (item.getMatchLabels() != null) {
+                    version.set(item.getMatchLabels().get(VERSION_LABEL));
+                }
+            }
+        }).build();
+        return Optional.ofNullable(version.get());
     }
 }

